### PR TITLE
Coop-Settle channel-state logic

### PR DESCRIPTION
**Pull request by ezdac**
### Implement the Coop-Settle logic built on-top of the Withdraw protocol.
This PR implements the logic as specified in [this](https://github.com/raiden-network/spec/issues/342) proposal.
It leverages the internal locking of channels with the `NettingChannelEndState.withdraws_pending` checks, that are
already in place for the withdraws. 
The most important additional verification for coop-settle is:
- we don't accept any coop-settle related withdraws when there are pending payments (locks) for us or the partner
- we don't emit withdraw-specific events (especially contract calls) from the state-transition for coop-settle related withdraw messages.

### This PR lacks:
- Changes to the withdraw-message schema
- new API endpoint or integration in existing API endpoint that would trigger a `ActionInitCoopSettle`
- processing of the `ContractSendCoopSettle` event
- thus no blockchain interaction is implemented
- Integration tests (since the full coop-settle lifecycle is not implemented)

### CODE REVIEW

File: raiden/tests/unit/channel_state/test_coop_settle.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+from raiden.transfer.state_change import (
+    ActionChannelCoopSettle,
+    ReceiveWithdrawConfirmation,
+    ReceiveWithdrawRequest,
+)
+from raiden.utils.packing import pack_withdraw
+from raiden.utils.signer import LocalSigner
+
+
+def make_receive_coop_settle_withdraw_confirmation(
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_coop_settle.py]: I don't think there's need to make this public. If it's used just within this one test module, please make it private.

File: raiden/tests/unit/channel_state/test_coop_settle.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+        sender=channel_state.partner_state.address,
+        participant=channel_state.our_state.address,
+        # pylint: enable=no-member
+        nonce=nonce,
+        expiration=withdraw_expiration,
+    )
+    return withdraw_confirmation
+
+
+def make_receive_coop_settle_withdraw_request(
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_coop_settle.py]: Similarly here.

File: raiden/tests/unit/channel_state/test_coop_settle.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        participant=channel_state.partner_state.address,
+        # pylint: enable=no-member
+        nonce=nonce,
+        expiration=withdraw_expiration,
+        coop_settle=True,
+    )
+    return withdraw_request
+
+
+def assert_coop_settle_state(
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_coop_settle.py]: And here.

File: raiden/transfer/channel.py [hunk 4]
```diff
@@ -2787,6 +2786,14 @@ def state_transition(
             block_number=block_number,
             block_hash=block_hash,
         )
+    elif type(state_change) == ActionChannelCoopSettle:
```
Comment (istankovic) [inline raiden/transfer/channel.py]: Shouldn't we rather use `is` for such type comparisons?

File: raiden/tests/unit/channel_state/test_channel_deposit.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
+# pylint: disable=too-many-locals,too-many-statements,too-many-lines
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_channel_deposit.py:1]: I like the split! :+1:

File: raiden/transfer/channel.py (hunk 4 above)
Comment (ezdac) [reply raiden/transfer/channel.py]: That would probably be better.
But since those type of checks are all over the place, I would rather not have this refactored in this PR, and even less so only checking with `is` for the new code path.

File: raiden/tests/unit/channel_state/test_coop_settle.py (hunk 1 above)
Comment (ezdac) [reply raiden/tests/unit/channel_state/test_coop_settle.py]: Nice, today I learned that `_methods` will not get imported if you do the `from foo import *`
Changed.

File: raiden/tests/unit/channel_state/test_coop_settle.py (hunk 2 above)
Comment (ezdac) [reply raiden/tests/unit/channel_state/test_coop_settle.py]: Done

File: raiden/tests/unit/channel_state/test_coop_settle.py (hunk 3 above)
Comment (ezdac) [reply raiden/tests/unit/channel_state/test_coop_settle.py]: Done

**Comment by fredo** [comment]
> new API endpoint or integration in existing API endpoint that would trigger a ActionInitCoopSettle

Did we agree on having a specific endpoint for it? Or couldn't we optimisticly try to coop settle and if not possible do the old settlement process?

**Comment by ezdac** [comment]
> Did we agree on having a specific endpoint for it? Or couldn't we optimisticly try to coop settle and if not possible do the old settlement process?

That's the plan - reuse the channel-close endpoint: there try a coop-settle first and if that expires go with the traditional close/settle cycle.

File: raiden/transfer/channel.py [hunk 6]
```diff
@@ -1184,10 +1184,22 @@ def get_max_withdraw_amount(
...
-    This will leave the channel without funds, when
+    This will leave the channel without funds, when this is withdrawn from the contract,
+    or pending as offchain-withdraw.
 
     """
-    return WithdrawAmount(sender.total_withdraw + get_balance(sender, receiver))
+    sender_transferred_amount = 0
+    receiver_transferred_amount = 0
+
+    if sender.balance_proof:
```
Comment (istankovic) [inline raiden/transfer/channel.py]: Please use explicit checks for `None`.

File: raiden/transfer/channel.py [hunk 7]
```diff
@@ -1184,10 +1184,22 @@ def get_max_withdraw_amount(
...
+        sender_transferred_amount = sender.balance_proof.transferred_amount
+
+    if receiver.balance_proof:
```
Comment (istankovic) [inline raiden/transfer/channel.py]: Please use explicit checks for `None`.

File: raiden/transfer/channel.py [hunk 8]
```diff
@@ -1184,10 +1184,22 @@ def get_max_withdraw_amount(
...
+        receiver_transferred_amount = receiver.balance_proof.transferred_amount
+
+    return WithdrawAmount(
+        sender.contract_balance - sender_transferred_amount + receiver_transferred_amount
+    )
```
Comment (istankovic) [inline raiden/transfer/channel.py]: I would like to see the changes to `transfer/channel.py` moved into a separate commit, since this does not seem to be a part of the straightforward test-split topic.

File: raiden/transfer/channel.py [hunk 9]
```diff
@@ -1152,6 +1180,16 @@ def get_balance(sender: NettingChannelEndState, receiver: NettingChannelEndState
     )
 
 
+def get_max_withdraw_amount(
+    sender: NettingChannelEndState, receiver: NettingChannelEndState
+) -> WithdrawAmount:
+    """Calculates the maximum "total_withdraw_amount" for a channel.
+    This will leave the channel without funds, when
+
```
Comment (istankovic) [inline raiden/transfer/channel.py:1196]: Incomplete comment.

File: raiden/transfer/channel.py [hunk 10]
```diff
@@ -1743,7 +1780,14 @@ def send_expired_withdraws(
...
 
         nonce = get_next_nonce(channel_state.our_state)
-        channel_state.our_state.nonce = nonce
+        coop_settle = channel_state.our_state.initiated_coop_settle
+        if coop_settle is not None:
+            if (
+                coop_settle.total_withdraw_participant == withdraw_state.total_withdraw
+                and coop_settle.expiration == withdraw_state.expiration
+            ):
+                # We also declare the coop-settle as expired, when we initated it
```
Comment (istankovic) [inline raiden/transfer/channel.py]: initated -> initiated

File: raiden/transfer/channel.py [hunk 11]
```diff
@@ -1743,7 +1780,14 @@ def send_expired_withdraws(
             break
 
         nonce = get_next_nonce(channel_state.our_state)
-        channel_state.our_state.nonce = nonce
+        coop_settle = channel_state.our_state.initiated_coop_settle
+        if coop_settle is not None:
+            if (
+                coop_settle.total_withdraw_participant == withdraw_state.total_withdraw
+                and coop_settle.expiration == withdraw_state.expiration
```
Comment (istankovic) [inline raiden/transfer/channel.py:1796]: Is the equality operator appropriate here when checking expiration?

File: raiden/transfer/channel.py [hunk 12]
```diff
@@ -1868,6 +1912,54 @@ def handle_action_close(
     return TransitionResult(channel_state, events)
 
 
+def handle_action_coop_settle(
+    channel_state: NettingChannelState,
+    action_coop_settle: ActionChannelCoopSettle,
+    pseudo_random_generator: random.Random,
+    block_number: BlockNumber,
+) -> TransitionResult[NettingChannelState]:
+    events: List[Event] = list()
```
Comment (istankovic) [inline raiden/transfer/channel.py]: We should rather use a literal here: `[]`.

File: raiden/transfer/channel.py [hunk 13]
```diff
@@ -1868,6 +1912,54 @@ def handle_action_close(
...
+
+    our_max_total_withdraw = get_max_withdraw_amount(
+        channel_state.our_state, channel_state.partner_state
+    )
+    partner_max_total_withdraw = get_max_withdraw_amount(
+        channel_state.partner_state, channel_state.our_state
+    )
```
Comment (istankovic) [inline raiden/transfer/channel.py:1952]: These only need to be computed if `valid_coop_settle` below is `True`.

File: raiden/transfer/channel.py [hunk 14]
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
...
+        nonce=withdraw_request.nonce,
+        expiration=withdraw_request.expiration,
+        recipient_metadata=withdraw_request.sender_metadata,
+    )
+    channel_state.partner_state.withdraws_pending[withdraw_state.total_withdraw] = withdraw_state
+    channel_state.partner_state.nonce = withdraw_request.nonce
+
+    our_initiated_coop_settle = channel_state.our_state.initiated_coop_settle
+
+    if our_initiated_coop_settle is not None or withdraw_request.coop_settle is True:
```
Comment (istankovic) [inline raiden/transfer/channel.py]: We should avoid checking booleans with `is`. Just use plain `or withdraw_request.coop_settle:`

File: raiden/transfer/channel.py [hunk 15]
```diff
@@ -1743,7 +1780,14 @@ def send_expired_withdraws(
             break
 
         nonce = get_next_nonce(channel_state.our_state)
-        channel_state.our_state.nonce = nonce
```
Comment (karlb) [inline raiden/transfer/channel.py:1747]: Why don't we need this line anymore?

File: raiden/transfer/channel.py (hunk 10 above)
Comment (karlb) [reply raiden/transfer/channel.py]: And I think it should be `if` instead of `when`.

File: raiden/transfer/channel.py [hunk 16]
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
...
+                    "Partner did not withdraw with maximum balance."
+                )
+            )
+        if get_number_of_pending_transfers(channel_state.partner_state) > 0:
+            return transition_result_invalid(SuccessOrError("Partner has pending transfers."))
+
+        if our_initiated_coop_settle is not None:
+            # There is a coop-settle inplace that we initiated
+            # and partner is communicating their total withdraw with us
+            if not our_initiated_coop_settle.expiration == withdraw_request.expiration:
```
Comment (istankovic) [inline raiden/transfer/channel.py]: We should use `!=` for comparison here.

File: raiden/transfer/channel.py [hunk 17]
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
...
+            )
+            if get_number_of_pending_transfers(channel_state.our_state) > 0:
+                return transition_result_invalid(
+                    SuccessOrError("Partner initiated coop-settle, but we have pending transfers.")
+                )
+            partner_initiated_coop_settle = CoopSettleState(
+                total_withdraw_participant=withdraw_request.total_withdraw,
+                total_withdraw_partner=our_max_total_withdraw,
+                expiration=withdraw_request.expiration,
+            )
```
Comment (istankovic) [inline raiden/transfer/channel.py:2152]: Should we have a sanity check for `withdraw_request.expiration`? What happens if the partner sends us a request with unreasonably long expiration?

File: raiden/transfer/channel.py (hunk 11 above)
Comment (karlb) [reply raiden/transfer/channel.py:1796]: Since we don't allow a coop settle and withdraws at the same time, this whole if statement should always be true, shouldn't it?

File: raiden/transfer/channel.py (hunk 13 above)
Comment (karlb) [reply raiden/transfer/channel.py:1952]: But to calculate `valid_coop_settle`we need `our_max_total_withdraw`. Your statement is still true about `partner_max_total_withdraw`, but this way both of the assignments are close together.

File: raiden/transfer/channel.py [hunk 18]
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
...
-                canonical_identifier=withdraw.canonical_identifier,
-                total_withdraw=withdraw.total_withdraw,
-                partner_signature=withdraw.signature,
-                expiration=withdraw.expiration,
-                triggered_by_block_hash=block_hash,
+        our_initiated_coop_settle = channel_state.our_state.initiated_coop_settle
+        partner_initiated_coop_settle = channel_state.partner_state.initiated_coop_settle
+        if our_initiated_coop_settle is not None:
+            # There is a coop settle inplace  that we initiated
+            assert not partner_initiated_coop_settle, "Only one party can initiate a coop settle"
```
Comment (istankovic) [inline raiden/transfer/channel.py]: We should check explicitly against `None`.

File: raiden/transfer/channel.py [hunk 19]
```diff
@@ -1924,91 +2019,200 @@ def handle_action_set_reveal_timeout(
...
+            # Normal withdraw
+            if withdraw_confirmation.expiration >= block_number - channel_state.reveal_timeout:
+                withdraw_on_chain = ContractSendChannelWithdraw(
+                    canonical_identifier=withdraw_confirmation.canonical_identifier,
+                    total_withdraw=withdraw_confirmation.total_withdraw,
+                    partner_signature=withdraw_confirmation.signature,
+                    expiration=withdraw_confirmation.expiration,
+                    triggered_by_block_hash=block_hash,
+                )
+                events.append(withdraw_on_chain)
```
Comment (istankovic) [inline raiden/transfer/channel.py:2232]: It would be nice to keep the old comment on this block.

File: raiden/transfer/events.py [hunk 20]
```diff
@@ -51,6 +51,7 @@ class SendWithdrawRequest(SendMessageEvent):
     participant: Address
     expiration: BlockExpiration
     nonce: Nonce
+    coop_settle: Optional[bool] = False
```
Comment (istankovic) [inline raiden/transfer/events.py]: Why is this optional?

File: raiden/transfer/events.py [hunk 21]
```diff
@@ -124,6 +125,40 @@ def __repr__(self) -> str:
...
 
+@dataclass(frozen=True)
+class ContractSendChannelCoopSettle(ContractSendEvent):
+    """Event emitted if node wants to withdraw from current channel balance."""
+
+    canonical_identifier: CanonicalIdentifier
+    our_total_withdraw: WithdrawAmount
+    partner_total_withdraw: WithdrawAmount
+    expiration: BlockExpiration
+    # This are the partner's signatures of the two withdraws.
```
Comment (istankovic) [inline raiden/transfer/events.py]: This -> These

File: raiden/transfer/channel.py [hunk 22]
```diff
@@ -914,32 +918,58 @@ def is_valid_unlock(
...
+) -> SuccessOrError:
+    return is_valid_total_withdraw(channel_state, withdraw.total_withdraw)
+
+
+def is_valid_action_coopsettle(
+    channel_state: NettingChannelState,
+    coop_settle: ActionChannelCoopSettle,  # pylint: disable=unused-argument
+    total_withdraw: WithdrawAmount,
+) -> SuccessOrError:
+    result = is_valid_total_withdraw(channel_state, total_withdraw)
```
Comment (karlb) [inline raiden/transfer/channel.py:959]: This will fail if we don't have anything to withdraw in our channel. It is not very useful to settle a channel if you have no balance left, but it might still be nice to allow it.

It is also not obvious why we check the total_withdraw for one participant but not the other.

File: raiden/transfer/events.py [hunk 23]
```diff
@@ -124,6 +125,40 @@ def __repr__(self) -> str:
...
+    @property
+    def token_network_address(self) -> TokenNetworkAddress:
+        return self.canonical_identifier.token_network_address
+
+    def __repr__(self) -> str:
+        return (
+            f"{self.__class__.__name__}< "
+            f"canonical_identifier: {self.canonical_identifier} "
+            f"our_total_withdraw: {self.our_total_withdraw} "
+            f"parner_total_withdraw: {self.our_total_withdraw} "
```
Comment (istankovic) [inline raiden/transfer/events.py]: Typo, should be
```
             f"partner_total_withdraw: {self.partner_total_withdraw} "
```

File: raiden/transfer/events.py [hunk 24]
```diff
@@ -124,6 +125,40 @@ def __repr__(self) -> str:
...
+            f"expiration: {self.expiration} "
+            f"partner_signature_withdraw: {to_hex(self.signature_partner_withdraw)} "
+            f"partner_signature_our_withdraw: {to_hex(self.signature_our_withdraw)} "
```
Comment (istankovic) [inline raiden/transfer/events.py]: I'd prefer to use field names as they are, instead of renaming them for the `__repr__` output.

File: raiden/transfer/state.py [hunk 25]
```diff
@@ -304,6 +305,15 @@ class PendingWithdrawState:
     recipient_metadata: Optional[AddressMetadata] = None
 
 
+@dataclass
+class CoopSettleState:
+    total_withdraw_participant: WithdrawAmount
+    total_withdraw_partner: WithdrawAmount
```
Comment (karlb) [inline raiden/transfer/state.py]: The usual naming pair is `our` (instead of `participant`) and `partner`.

File: raiden/transfer/state_change.py [hunk 26]
```diff
@@ -116,6 +116,18 @@ def channel_identifier(self) -> ChannelID:
         return self.canonical_identifier.channel_identifier
 
 
+@dataclass(frozen=True)
+class ActionChannelCoopSettle(StateChange):
+    """Withdraw funds from channel."""
```
Comment (karlb) [inline raiden/transfer/state_change.py]: This comment is copy-pasted from `ActionChannelWithdraw` and needs updating.

File: raiden/transfer/events.py (hunk 20 above)
Comment (karlb) [reply raiden/transfer/events.py]: Because not all withdraws are coop settlements. There is still the "normal" withdraw.

File: raiden/transfer/state.py (hunk 25 above)
Comment (karlb) [reply raiden/transfer/state.py]: I see now that these are not "our" and "our partner" but rather the participant initiating the coop settle and the participant's partner. If we don't find better names, we should at least add a short comment.

File: raiden/transfer/channel.py (hunk 6 above)
Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 7 above)
Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 8 above)
Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 10 above)
Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 11 above)
Comment (istankovic) [reply raiden/transfer/channel.py:1796]: Hmm... not sure about that, but looking at it the other way, I would say it seems ok to reset the `coop_settle` if _any_ withdraw expired. So I guess it could be simplified.

However, I don't like doing channel state changes here, this seems rather like a function that is supposed to return events, not cause side-effects like that. @ulope Any ideas for a better place?

File: raiden/transfer/channel.py (hunk 12 above)
Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 13 above)
Comment (istankovic) [reply raiden/transfer/channel.py:1952]: Ah, right.

File: raiden/transfer/channel.py (hunk 14 above)
Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 16 above)
Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 18 above)
Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 19 above)
Comment (istankovic) [reply raiden/transfer/channel.py:2232]: Done.

File: raiden/transfer/events.py (hunk 23 above)
Comment (istankovic) [reply raiden/transfer/events.py]: Ah, fixed in a later commit.

File: raiden/transfer/events.py (hunk 24 above)
Comment (istankovic) [reply raiden/transfer/events.py]: Changed to original field names.

File: raiden/transfer/state.py (hunk 25 above)
Comment (istankovic) [reply raiden/transfer/state.py]: Maybe... `total_withdraw_initiator`?

Comment (istankovic) [reply raiden/transfer/state.py]: Went with the initiator and left the other as partner.

File: raiden/transfer/state_change.py (hunk 26 above)
Comment (istankovic) [reply raiden/transfer/state_change.py]: Done.

File: raiden/transfer/channel.py (hunk 9 above)
Comment (istankovic) [reply raiden/transfer/channel.py:1196]: Fixed in a later commit.

File: raiden/tests/unit/channel_state/test_coop_settle.py [hunk 27]
```diff
@@ -0,0 +1,448 @@
...
+from raiden.utils.packing import pack_withdraw
+from raiden.utils.signer import LocalSigner
+
+
+def _make_receive_coop_settle_withdraw_confirmation(
+    channel_state, pseudo_random_generator, signer, nonce, total_withdraw, withdraw_expiration
+):
+    packed = pack_withdraw(
+        canonical_identifier=channel_state.canonical_identifier,
+        # pylint: disable=no-member
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_coop_settle.py]: This seems unnecessary, at least with my version of pylint.

File: raiden/tests/unit/channel_state/test_coop_settle.py [hunk 28]
```diff
@@ -0,0 +1,448 @@
+# pylint: disable=too-many-locals,too-many-statements,too-many-lines
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_coop_settle.py]: This seems unnecessary, at least with my version of pylint.

File: raiden/tests/unit/channel_state/test_coop_settle.py (hunk 27 above)
Comment (istankovic) [reply raiden/tests/unit/channel_state/test_coop_settle.py]: Removed this and the rest like it.

File: raiden/tests/unit/channel_state/test_coop_settle.py (hunk 28 above)
Comment (istankovic) [reply raiden/tests/unit/channel_state/test_coop_settle.py]: Removed.

File: raiden/tests/unit/channel_state/test_coop_settle.py [hunk 29]
```diff
@@ -0,0 +1,448 @@
...
+        channel_state.our_state,
+        channel_state.partner_state,
+        our_initial_model=our_model1,
+        our_current_nonce=1,
+        coop_settle_expiration=expiration,
+        initiated_coop_settle=True,
+    )
+    # The partner still has full balance in his state.
+    # He could therefore initiate a transfer in our direction now and we would accept it
+    # The then could try to continue the coop-settle with old balances.
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_coop_settle.py]: Typo: "The then"

File: raiden/tests/unit/channel_state/test_coop_settle.py [hunk 30]
```diff
@@ -0,0 +1,448 @@
...
+    # partner's view
+    _assert_coop_settle_state(
+        channel_state.our_state,
+        channel_state.partner_state,
+        our_initial_model=our_model1,
+        our_current_nonce=2,
+        coop_settle_expiration=expiration,
+        initiated_coop_settle=False,
+    )
+
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_coop_settle.py:266]: The comments here need to be swapped -- "our view" should be where "partner's view" is and vice versa.

Comment (istankovic) [reply raiden/tests/unit/channel_state/test_coop_settle.py:266]: Done.

File: raiden/tests/unit/channel_state/test_coop_settle.py [hunk 31]
```diff
@@ -0,0 +1,448 @@
...
+    assert len(iteration.events) == 1
+
+    # our view
+    _assert_coop_settle_state(
+        channel_state.our_state,
+        channel_state.partner_state,
+        our_initial_model=our_model1,
+        our_current_nonce=1,
+        coop_settle_expiration=expiration,
+        initiated_coop_settle=True,
```
Comment (istankovic) [inline raiden/tests/unit/channel_state/test_coop_settle.py:323]: This is not clear to me. We're setting `initiated_coop_settle` to `True`, even though it was the partner who sent the coop settle withdraw request. Could this be because we set the `CoopSettleState` manually? I'm confused.

File: raiden/tests/unit/channel_state/test_coop_settle.py (hunk 29 above)
Comment (istankovic) [reply raiden/tests/unit/channel_state/test_coop_settle.py]: Fixed.

File: raiden/transfer/channel.py (hunk 15 above)
Comment (ulope) [reply raiden/transfer/channel.py:1747]: Good question. 
[This is the commit where it was introduced](https://github.com/raiden-network/raiden/commit/17e5ae45a147990e45ae471065ce8beb4e596a70)

File: raiden/transfer/channel.py (hunk 11 above)
Comment (ulope) [reply raiden/transfer/channel.py:1796]: All these functions in here are in the end (through various layers) called from `raiden.transfer.channel.state_transition`. 
Also if you check the other functions, they all do direct state manipulation. So this is fine.

> Is the equality operator appropriate here when checking expiration?

I assume this is intended to check that the expiring withdraw is the one that belongs to the coop settle and not some other unrelated withdraw (which as @karlb said shouldn't be possible).

File: raiden/transfer/channel.py (hunk 17 above)
Comment (ulope) [reply raiden/transfer/channel.py:2152]: There seems to be little incentive for the peer to do that.
Choosing a very long expiration locks funds for this period. If the peer is the one requesting the coop settle it stands to reason that they are the one with balance to gain.

Also in the worst case either party can always claim on-chain.

Some previous discussions: 
- https://github.com/raiden-network/raiden/issues/864
- https://github.com/raiden-network/raiden/pull/4289

File: raiden/transfer/channel.py (hunk 4 above)
Comment (ulope) [reply raiden/transfer/channel.py]: This would be a good candidate to also refactor with `functools.singledispatch` (c.f. #7252) (in another PR)

File: raiden/transfer/events.py (hunk 20 above)
Comment (ulope) [reply raiden/transfer/events.py]: But what's the third state? 
`Optional[bool]` is tri-state (`True | False | None`).

I don't see `None` being used and also can't think what that would mean in this context.

File: raiden/transfer/channel.py [hunk 32]
```diff
@@ -1184,10 +1184,22 @@ def get_max_withdraw_amount(
     sender: NettingChannelEndState, receiver: NettingChannelEndState
 ) -> WithdrawAmount:
     """Calculates the maximum "total_withdraw_amount" for a channel.
-    This will leave the channel without funds, when
+    This will leave the channel without funds, when this is withdrawn from the contract,
+    or pending as offchain-withdraw.
 
     """
-    return WithdrawAmount(sender.total_withdraw + get_balance(sender, receiver))
```
Comment (ulope) [inline raiden/transfer/channel.py]: This doesn't seem right. Why would we not need to include the previous off-chain withdraws in the calculation (which is what the `get_balance()` function does)?

Comment (ulope) [reply raiden/transfer/channel.py]: Ok, actually this is correct since coop settle requires no pending withdraws.
However I still think it would be nicer to use `get_balance` instead of ad-hoc reimplementing half of it here.

Comment (istankovic) [reply raiden/transfer/channel.py]: Done.

File: raiden/transfer/channel.py (hunk 15 above)
Comment (istankovic) [reply raiden/transfer/channel.py:1747]: I brought back the line, it's my impression that we need to update the nonce if we're to emit events that use it.

File: raiden/transfer/events.py (hunk 20 above)
Comment (istankovic) [reply raiden/transfer/events.py]: I made it non-optional.

File: raiden/transfer/channel.py (hunk 22 above)
Comment (ulope) [reply raiden/transfer/channel.py:959]: To me it looks like `is_valid_total_withdraw` checks both sides.

**Review by istankovic — APPROVED** [review]
